automata/regex/parser.py

```diff
@@ -10,7 +10,7 @@
                                     parse_postfix_tokens, tokens_to_postfix,
                                     validate_tokens)
 
-RESERVED_CHARACTERS = frozenset({'*', '|', '(', ')', '?', ' ', '\t', '&', '+', '.'})
+RESERVED_CHARACTERS = frozenset({'*', '|', '(', ')', '?', ' ', '\t', '&', '+', '.', '@'})
```

Comment:
Tbh not sure about this character. '@' is quite used for user data (username or e-mail for example). I think that either we introduce it in a different version, or we enable some form of configuration for the parser...

I would vote for the latter because this operator is more exotic and from an API design perspective, it makes sense to make it easy to perform the common tasks. So if someone is using the shuffle operator, he/she is probably fine with passing around a few extra parameters

Reply:
You raise a good point about using the `@` symbol, although I'm not so sure about having the user pass in a character every time to be used. I think this would introduce the possibility of more confusion, along with having to pass through a bunch of options to the parser itself.

Maybe the percent sign `%` is less intrusive? Looking at regex syntax in other places, this symbol doesn't seem to be used anywhere (see [here](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX-Extended_Regular_Expressions) and [here](https://www.brics.dk/automaton/doc/dk/brics/automaton/RegExp.html)). @caleb531 what are your thoughts?

Reply:
@eliotwrobson Well, because the percent sign is often used in URL escape sequences (e.g. `%20`), I would prefer to avoid it.

I personally liked @EduardoGoulart1's suggestion of `~`; even if it doesn't _look_ like Cyrillic sha, the waviness of `~` reminds me of shuffling (at least in my strange mind lol). As long as we document what the symbol is, it shouldn't matter if it doesn't look like sha.

Reply:
Also, as we add more symbols of special meaning to the regex grammar, I am thinking it's about time to add an `escape` function (much like `re.escape` in the stdlib) which properly escapes the appropriate literal characters (e.g. `someperson\+suffix1@gmail\.com`). If there's no way of escaping regex symbols currently, we should add this (preferably in a separate PR to recognize backslashes in an automaton regex).

Reply:
@caleb531 that makes sense. I shied away from the `~` because the [brics java automata](https://www.brics.dk/automaton/doc/dk/brics/automaton/RegExp.html) library uses it for compliment, but I don't think it's a great idea to implement that operation for regexs. I do agree though that the `~` seems to invoke feelings of interleaving.

Currently, only alphanumeric characters are supported in the regex parsing here (which I'm now realizing is kindof a breaking change from v6, whoops). But I think some way of adding escape characters could be useful to people. I think it's just a matter of reconfiguring the lexer to treat characters coming after a backslash as a literal. But I think it should definitely be in a separate PR.

Reply:
@eliotwrobson Oh yeah, forgot about `~` as complement. In that case, I think I'd actually prefer to eliminate `~` as a candidate symbol, as well. Otherwise, it might give the wrong impression (like a negative lookahead or something). Even though complement isn't useful in a regex context, it seems like the regex operators have equivalent NFA operations as well (e.g. regex `*` maps to NFA Kleene star).

Maybe this "trying to find the right character" debate is an indication that we're pressing against the limit of what new operations should be part of the regex grammar anyway. I mean, sure, the ASCII constraint is potentially arbitrary (i.e. if we just used sha, then we wouldn't need to have this conversation). But the shuffle operator is really just syntactic sugar for something that can technically already be expressed as a series of `|`'s.

I never learned about a shuffle operation in a regex context, so I'm just trusting both you guys that it's not only practical, but coherently fits within the theory as well

I saw you _just_ pushed a commit to use `~`, but given how I feel above, I'm no longer sure that's the right operator. But what do you think about `^` as the symbol to use? Kinda has a combinatoric feel to it (because, well, `^` is often used for exponentiation).

Reply:
@caleb531 I like the `^` as well. With respect to your other comment, it's actually pretty difficult to replicate the shuffle operation using just union `|` operators (as there's a combinatorial blowup in the number of terms needed), so adding support for the shuffle could actually be very convenient.

With respect to the larger vision, I actually kinda have the opposite viewpoint. A lot of other regex libraries are constrained by the fact that they use more efficient matching algorithms, so with this regex engine being oriented around NFA construction and available operations, I think it helps the library stand out to have these unique operations.

From a theory standpoint, the way to view these operations is being equivalent to operations on NFAs, and regex is just a shorthand syntax for this construction and operation. In that sense, any operation that can be performed with one or more NFAs makes sense theoretically, and I think it's cool that we have a chance to incorporate that here, while these ideas may not fit in as nicely with a more application-oriented regex implementation.

Reply:
@caleb531 No worries! I hope the discussion was as insightful for you as for me. And not a problem, the initial symbol change was just a find and replace, which is why I was able to push the changes so quickly 😆

Reply:
Symbol has been changed to `^`!

Reply:
@caleb531 @eliotwrobson I can't react so quickly to PRs, so sorry if I arrive late in the discussions

I just saw your comment that we only support alphanumeric characters. That basically means that my fear of breaking existing code was nonsense because the current code does not support `@` anyways. For example, `NFA.from_regex('.*@email.com')` throws a `LexerError: Invalid character '@'`

I'm still getting into the codebase, still I should have checked for that first